Add root-level package.json and set up workspaces.

datu925 opened the pull request:
We can go further if we need to, but this gives us the whole-project capabilities that I was looking for. It uses npm workspaces which are documented here: https://docs.npmjs.com/cli/v7/using-npm/workspaces. I also found this [blog post](https://www.pixelmatters.com/blog/how-to-manage-multiple-front-end-projects-with-a-monorepo) helpful.

In order for this to work, I added a package.json at the project root, and then I had to delete all package-lock.json files from the rest of the project. But you are allowed to keep using your own package.json for better modularity. In general, install dependencies in your own package.json files and not the root one. The only dependency I included at root was ts-node to make it easy to run scripts from root since the other commands require you to be in root as well.

To compile all sub-projects: `npm exec --ws -- tsc`
To run tests (or other scripts) for all sub-projects: `npm run test --workspaces --if-present`

@cdcyang I made some changes to your tsconfig file to write the compiled JS to a `build/` directory – not sure how you were compiling/running things before, but this is what RA does and it keeps things tidy. I also removed the dummy `test` script since it was causing errors (as intended, but it was starting to get annoying).

cdcyang commented on .gitignore:
```diff
@@ -143,6 +143,7 @@ __pycache__
 
 # Compiled JavaScript
 pipeline/build/
+backend/build/
```
Can these two be replaced by `**/build/`?

datu925 replied:
I did this in a different PR, so yes, this is `**/build` now.

datu925 commented:
> > We can go further if we need to, but this gives us the whole-project capabilities that I was looking for. It uses npm workspaces which are documented here: https://docs.npmjs.com/cli/v7/using-npm/workspaces. I also found this [blog post](https://www.pixelmatters.com/blog/how-to-manage-multiple-front-end-projects-with-a-monorepo) helpful.
> > In order for this to work, I added a package.json at the project root, and then I had to delete all package-lock.json files from the rest of the project. But you are allowed to keep using your own package.json for better modularity. In general, install dependencies in your own package.json files and not the root one. The only dependency I included at root was ts-node to make it easy to run scripts from root since the other commands require you to be in root as well.
> > To compile all sub-projects: `npm exec --ws -- tsc` To run tests (or other scripts) for all sub-projects: `npm run test --workspaces --if-present`
> > @cdcyang I made some changes to your tsconfig file to write the compiled JS to a `build/` directory – not sure how you were compiling/running things before, but this is what RA does and it keeps things tidy. I also removed the dummy `test` script since it was causing errors (as intended, but it was starting to get annoying).
> 
> As vercel only permits serverless functions on the backend (https://vercel.com/templates/other/fastify-serverless-function), I don't build the backend project locally, but instead deploy to a dev instance of vercel using the `vercel dev` command to verify that the project works as expected.
> 
> I have checked that the backend still function the same as before and I am happy for this change to merge as long as the front-end deployment is fixed :)

I think we ran into this. https://stackoverflow.com/questions/70117752/unable-to-deploy-a-next-js-monorepo-using-workspaces-to-vercel. I'll try it with yarn and request review again when it's done.

datu925 commented:
@cdcyang @rickycheah this change is ready for review now.

You can run commands from root for your backend like this:
`yarn workspace backend <name of script>`

To do things across all workspaces, you run `yarn workspaces foreach -A` (the -A flag is for --all):
`yarn workspaces foreach -A run tsc`
`yarn workspaces foreach -A run test`


This change includes switching over to Yarn 2 (aka Yarn modern) from npm. Migration should be relatively straightforward with https://yarnpkg.com/migration/guide. It's just the first few steps. But I want to make sure this is okay before springing it on you.
